library/src/scala/internal/quoted/PickledSplices.scala

```diff
@@ -0,0 +1,23 @@
+package scala.internal.quoted
+
+import scala.quoted._
+
+/** Provider of expressions and types that will fill the holes a pickled quote */
+trait PickledSplices:
+
+  /** Expression that will fill the hole `Hole(<idx> | <args>*)` */
+  def exprSplice(idx: Int)(args: Seq[Any /* Expr[Any] | Type[?] */])(qctx: QuoteContext): Expr[Any]
+
+  /** Type that will fill the hole `Hole(<idx> | <args>*)` */
+  def typeSplice(idx: Int)(args: Seq[Any /* Expr[Any] | Type[?] */]): Type[?]
```

Comment:
Conceptually, it seems to be part of the compiler interface. Will it be better to put it there?

Reply:
Not so sure. This represents a quote that has not been contextualized yet. But maybe the `PickledSplices` abstraction can be hidden with the `PickledExpr` and `PickledType`.

Reply:
This seems to be the cleanest option

Reply:
It seems I miss some context here. Could you elaborate a little bit about what kind of compiler changes that this PR aims to defend against and how this PR achieves that goal?

Reply:
The idea is to be able to change the code that is generated to fill the holes of the pickled quote. Right now we use a sequence of lambdas to keep these and we cannot easily change it in future versions because this leaks into the API. One possible improvement is [described here](https://github.com/lampepfl/dotty/pull/9948/files#diff-0cf7b44467678a539d53ee72a7089228aa9b2684654e8139222243a1bc286913R15-R17).

With this API we could then evolve the implementation of this code without breaking the old encodings.
```scala
// Code generated by compiller
PickledSplices.make(Seq(
  (args: Seq[Any]) => (qctx: QuoteContext) => ..., // Some Expr
  (args: Seq[Any]) => (qctx: QuoteContext) => ..., // Another Expr
))
// could be replaced without breaking API compatibility by a more performant version
new PickledSplices {
  def exprSplice(idx: Int)(args: Seq[Any])(qctx: QuoteContext): Expr[Any] = 
    idx match 
      case 0 => ... // Some Expr
      case 1 => ... // Another Expr
}
```
In this case, we would be avoiding allocations of the sequence and the lambdas. But other unforeseen improvements could arise in the future.

On the other side, `PickledExpr` and `PickledType` aim to be able to change the binary encoding of the pickled quote. Now we convert the `Array[Byte]` into a string using the `Base64` encoder, but we should be able to change this encoding without breaking existing code.

Reply:
Thanks for the explanation 👍

Reply:
Minor: another possibility is to just define a single trait to hold all the protocols instead of defining 3 different traits.

Reply:
I added an extra commit where I refactored it in a single interface.